pymc/sampling_jax.py

```diff
@@ -95,6 +97,39 @@ def logp_fn_wrap(x):
     return logp_fn_wrap
 
 
+# Adopted from arviz numpyro extractor
+def _sample_stats_to_xarray(posterior):
+    """Extract sample_stats from NumPyro posterior."""
+    rename_key = {
+        "potential_energy": "lp",
+        "adapt_state.step_size": "step_size",
+        "num_steps": "n_steps",
+        "accept_prob": "acceptance_rate",
+    }
+    data = {}
+    for stat, value in posterior.get_extra_fields(group_by_chain=True).items():
+        if isinstance(value, (dict, tuple)):
+            continue
+        name = rename_key.get(stat, stat)
+        value = value.copy()
+        data[name] = value
+        if stat == "num_steps":
+            data["tree_depth"] = np.log2(value).astype(int) + 1
+    return data
+
+
+def _get_log_likelihood(model, samples):
+    "Compute log-likelihood for all observations"
+    data = {}
+    for v in model.observed_RVs:
+        logp_v = replace_shared_variables([logpt(v)])
+        fgraph = FunctionGraph(model.value_vars, logp_v, clone=False)
+        jax_fn = jax_funcify(fgraph)
+        result = jax.vmap(jax.vmap(jax_fn))(*samples)[0]
```

Comment:
Out of curiosity, would we expect any benefits to jit_compiling this outer `vmap`?

Reply:
Would it be possible to use a similar approach with Aesara directly? 

Here we only loop over observed variables in order to get the pointwise log likelihood. We had some discussion about this in https://github.com/pymc-devs/pymc/pull/4489 but ended up keeping the 3 nested loops over variables, chains and draws.

Reply:
It should be possible, but requires a Aesara Scan, and at least for small models this was not faster that python looping when I checked it. Here is a Notebook that documents some things I tried: https://gist.github.com/ricardoV94/6089a8c46a0e19665f01c79ea04e1cb2

It might be faster if using shared variables...

Reply:
No idea. I think the easiest thing to do is just benchmark it. I don't even call `optimize_graph` on either the graph in this function or the main sample routine.

When I change

`result = jax.vmap(jax.vmap(jax_fn))(*samples)[0]` to 
`result = jax.jit(jax.vmap(jax.vmap(jax_fn)))(*samples)[0]`

I don't really get a speed-up until there are millions of samples.

Reply:
> I  don't even call `optimize_graph` on either the graph in this function or the main sample routine

We should definitely call `optimize`, otherwise the computed logps may not correspond to the ones used during sampling. For instance we have many optimizations that improve numerically stability, so you might get underflows to `-inf` for samples that were accepted, which could screw up things downstream.

Reply:
> It should be possible, but requires a Aesara Scan, and at least for small models this was not faster than python looping when I checked it.

Then it's probably not worth it. I was under the impression it would be possible to vectorize/broadcast the operation from the conversations in #4489 and in slack.

Reply:
It must be possible, since the vmap above work just fine. I just have no idea how they do it xD, or how/if you could do it in Aesara

Reply:
Alright. I'm going to make a separate PR for some this other stuff.

Reply:
Cool, feel free to tag me if you want me to review, I am not watching PRs. I can already say I won't be able to help with the vectorized log_likelihood thing, I tried and I lost much more time with that than what would have been healthy. I should be able to help with coords and dims though